File: js/src/2265-averageOfSubtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * 判断当前节点的值等于子树的值的平均值，如果是小数则向下取整
 * https://leetcode.cn/problems/count-nodes-equal-to-average-of-subtree/submissions/
 * @param {TreeNode} root
 * @return {number}
 * 84 ms, 在所有 JavaScript 提交中击败了25.00%
 */
const averageOfSubtree = function(root) {
  const runNode = (node) => {
    let sum = node.val;
    let number = 1;
    // 叶子结点一定满足
    if (!node.left && !node.right) {
      result++;
      return { sum, number };
    }
    // 有子节点，那么需要判断
    if (node.left) {
      const { sum: sum1, number: number1 } = runNode(node.left);
      sum += sum1;
      number += number1;
    }
    if (node.right) {
      const { sum: sum2, number: number2 } = runNode(node.right);
      sum += sum2;
      number += number2;
    }
    // 判断当前子树是否满足条件
    if (Math.floor(sum / number) === node.val) {
      result++;
    }
    return { sum, number };
  };
  let result = 0;
  if (root) {
    runNode(root);
  }
  return result;
};

export { averageOfSubtree };

File: python/src/2265-averageOfSubtree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        # 辅助函数 dfs，判断节点对应的子树是否满足，返回子树的和和节点个数
        def run_node(root: Optional[TreeNode]):
            # nonlocal 在函数内部创建一个函数，该函数使用变量 result 作为非局部变量
            nonlocal result
            if not root:
                return 0, 0
            left_node_result = run_node(root.left)
            right_node_result = run_node(root.right)
            value = left_node_result[0] + right_node_result[0] + root.val
            count = left_node_result[1] + right_node_result[1] + 1
            # Python 中两个斜杠即双斜杠 //表示地板除，即先做除法（/），然后向下取整（floor），等于 math.floor(value/count)
            if root.val == value // count:
                result += 1
            return value, count

        result = 0
        run_node(root)
        return result
